### ISSUE TITLE
xref with keydef generating wrong link

### ISSUE THREAD

**Issue by gabrielmcg** [dita-ot/dita-ot#2250]
I created a keydef for an image: 

`<keydef keys="caution_image" 
        href="../media/images/caution.png"
        format="png"/>
`

and then referenced it in content:

`<p><image keyref="caution_image"/></p>`  
 `<p><xref keyref="caution_image">Download</xref></p>
`

The url for the image is created correctly:

`<p class="p"><img class="image" src="../media/images/caution.png" /></p>`

but the url for the xref doesn't have the correct relative url:

`<p class="p"><a class="xref" href="media/images/caution.png">Download</a></p>`

Is this a bug in the toolkit? or is the source wrong? I have added the files below - I tested with 2.2.2
# 

[dita-bug.zip](https://github.com/dita-ot/dita-ot/files/161574/dita-bug.zip)

**Comment by eerohele** [comment]
Sounds very much like what @shaneataylor describes [here](https://github.com/dita-ot/dita-ot/issues/1951#issuecomment-174681131).

**Comment by drmacro** [comment]
Note that having a cross reference to an image doesn't make any sense--an image is not a thing you can usefully navigate to.

**Comment by eerohele** [comment]
>Note that having a cross reference to an image doesn't make any sense--an image is not a thing you can usefully navigate to.

You could have a cross-reference to something like a PDF file, though, and this bug affects that scenario as well.

There's multiple issues here. One is that keydefs that point to non-external non-DITA resources aren't resolved correctly. Tacking something like this on top of #2223 should fix the issue (untested — this is more like a reminder-to-self type of thing):

```diff
diff --git a/src/main/java/org/dita/dost/writer/KeyrefPaser.java b/src/main/java/org/dita/dost/writer/KeyrefPaser.java
index e344070..d47652f 100644
--- a/src/main/java/org/dita/dost/writer/KeyrefPaser.java
+++ b/src/main/java/org/dita/dost/writer/KeyrefPaser.java
@@ -402,6 +402,12 @@ public final class KeyrefPaser extends AbstractXMLFilter {
                                     final URI f = currentFile.resolve(targetOutput);
                                     normalProcessingRoleTargets.add(f);
                                 }
+                            } else if (!ATTR_FORMAT_VALUE_DITA.equals(elem.getAttribute(ATTRIBUTE_NAME_FORMAT))
+                                    && !ATTR_SCOPE_VALUE_EXTERNAL.equals(elem.getAttribute(ATTRIBUTE_NAME_SCOPE))) {
+                                valid = true;
+                                final File topicFile = toFile(currentFile.resolve(stripFragment(keyDef.source.resolve(href))));
+                                final URI targetOutput = URLUtils.getRelativePath(currentFile, topicFile.toURI());
+                                XMLUtils.addOrSetAttribute(resAtts, refAttr, targetOutput.toString());
                             } else {
                                 valid = true;
                                 final URI targetOutput = normalizeHrefValue(href, elementId);
```

Another is that such keydefs are added onto the map/topic list for some reason:

```
λ : tail temp/.job.xml
    <string>false</string>
  </property>
  <files>
    <file src="file:/Users/eerohele/Git/dita-ot-issues/fixtures/1951/media/images/caution.png" uri="media/images/caution.png" path="media/images/caution.png" format="dita" resource-only="true"/>
    <file src="file:/Users/eerohele/Git/dita-ot-issues/fixtures/1951/root.ditamap" uri="root.ditamap" path="root.ditamap" format="ditamap"/>
    <file src="file:/Users/eerohele/Git/dita-ot-issues/fixtures/1951/content/test.dita" uri="content/test.dita" path="content/test.dita" format="dita" has-keyref="true" non-conref-target="true" target="true"/>
    <file src="file:/Users/eerohele/Git/dita-ot-issues/fixtures/1951/defs/defs.ditamap" uri="defs/defs.ditamap" path="defs/defs.ditamap" format="ditamap" resource-only="true"/>
  </files>
</job>
```

Which will cause errors in the `debug-filter` stage:

```
     [java] debug-filter:
     [java]    [filter] Using Xerces grammar pool for DTD and schema caching.
     [java]    [filter] Processing file:/Users/eerohele/Git/dita-ot-issues/fixtures/1951/media/images/caution.png
     [java]    [filter] Invalid byte 1 of 1-byte UTF-8 sequence.
```

Which in turn will cause the image (or other referenced file) to be corrupted in the output. Didn't get to the bottom of that yet, though.

**Comment by mironovalexey** [comment]
Probably resolved in https://github.com/dita-ot/dita-ot/commit/3612e55aab141e664a78c2720e65b8a5759eae9b

**Comment by robander** [comment]
Fixed with #3234
